docs/layer_reference/loss.rst

```diff
@@ -4,6 +4,13 @@
 Loss Functions
 ==============
 
+LossLayer
```

Comment:
This is maybe misleading here? The loss layer is not a loss but a layer.

Reply:
I did not know of a better location, but I will definitely add some text at the header later on. Maybe it gets clear then.

Reply:
Also, btw, this is an example of a layer which I want to remove or deprecate as well -- or rather, I want to make all other losses just normal layers, and then this becomes obsolete. (We discussed this already.)

So I'm not sure if I would mention it in the doc. Or at least, it should have this disclaimer.

Reply:
I will add the disclaimer then. The loss layer should be part of the docs, as some people asked about how to get losses inside your network.

Reply:
Well, yea, but then, I think I would not tell them to use `LossLayer`, but instead to just calculate the loss directly, whatever it is. We have all the basic layers to calculate all kinds of losses. Or you could use `EvalLayer`. This is always better (more clear, more straight-forward, not going to be deprecated).

Reply:
But then you would have to deal with the mask explicitly. Your suggestion sounds like a lot of overhead, and leaves a lot of room for possible mistakes when you want to replicate an existing loss.

Reply:
> But then you would have to deal with the mask explicitly.

No. Why? All layers in RETURNN handle that automatically.

> Your suggestion sounds like a lot of overhead, and leaves a lot of room for possible mistakes when you want to replicate an existing loss.

No? I think this becomes easier, shorter, more straight forward. Using the `LossLayer` on the other side is kind of unintuitive.

Reply:
I want to make sure I understand:
You are suggesting to not use something like
`"output": {..., "loss": "mse", "target": "classes"}`
and not something like
`"mse_layer": {"class": "loss", "loss_": "mse", "target_": "classes", "loss": "as_is", "from": "output"}`
but instead
`"mse_layer": {"class": "eval", "eval": "tf.reduce_mean((source(0) - source(1)) ** 2)", "from": ["output", "data:classes"], "loss": "as_is"}` ?

Reply:
> I want to make sure I understand:
> You are suggesting to not use something like
> `"output": {..., "loss": "mse", "target": "classes"}`

No, I'm not talking about this in this context here. This is a separate topic, which we discussed elsewhere. But this is related, of course. And you are right, this is sth I don't like too much, to have the distinction between losses and layers. I discuss this in the document about the new RETURNN interface.

> and not something like
> `"mse_layer": {"class": "loss", "loss_": "mse", "target_": "classes", "loss": "as_is", "from": "output"}`

There would not be `loss: as_is` here (or maybe it can be, but this is not what we discuss here), but rather we just speak about how to calculate a loss, and use it elsewhere. I.e. what the `LossLayer` is for.

But yes, I recommend against using `LossLayer`, because I think this is quite unintuitive. To understand what this layer returns, and what it actually does, you need to understand quite some internals of RETURNN.

> but instead
> `"mse_layer": {"class": "eval", "eval": "tf.reduce_mean((source(0) - source(1)) ** 2)", "from": ["output", "data:classes"], "loss": "as_is"}` ?

This is not the equivalent of above (which kind of proves the point that `LossLayer` is unintuitive), and this is definitely what you should never do. Specifically, you never should use any of the `reduce` functions in an eval layer. We have the `ReduceLayer` for that. But also, the `LossLayer` would not reduce.
And again, remove the `loss: as_is` here.

The equivalent layer, and what I suggest what you should do instead, is e.g.:
```
"mse_layer": {"class": "eval", "eval": "(source(0) - source(1)) ** 2)", "from": ["output", "data:classes"]}
```
It's just much more straight-forward and clear what this does.

In the future, I also would like to reduce the usage of `EvalLayer`, and just use basic layers instead. But this is maybe too much overkill here, although only because of the syntax. With the new Python syntax, you directly would write:
```
mse_layer = (output - classes) ** 2.
```

Btw, your name `mse_layer` is misleading, as this does in fact not do the "mean".

Btw, if you then use `loss: as_is`, i.e. the `AsIsLoss` on such an output, the `AsIsLoss` (as all other losses) takes care of the reduction (mean or sum, depending how you configure it).

Reply:
> But yes, I recommend against using `LossLayer`, because I think this is quite unintuitive. To understand what this layer returns, and what it actually does, you need to understand quite some internals of RETURNN.
> 
> > but instead
> > `"mse_layer": {"class": "eval", "eval": "tf.reduce_mean((source(0) - source(1)) ** 2)", "from": ["output", "data:classes"], "loss": "as_is"}` ?
> 
> This is not the equivalent of above (which kind of proves the point that `LossLayer` is unintuitive), and this is definitely what you should never do. Specifically, you never should use any of the `reduce` functions in an eval layer. We have the `ReduceLayer` for that. But also, the `LossLayer` would not reduce.
> And again, remove the `loss: as_is` here.
> 
> The equivalent layer, and what I suggest what you should do instead, is e.g.:
> 
> ```
> "mse_layer": {"class": "eval", "eval": "(source(0) - source(1)) ** 2", "from": ["output", "data:classes"]}
> ```
> 
> It's just much more straight-forward and clear what this does.

Ah now I understand the Problem, the `LossLayer` is just not doing what anyone would think it does (But this could be fixed with a correct docstring). 

Well, I will add your explanation to the note in the docs and recommend to not use the `LossLayer` at all.

Reply:
> Ah now I understand the Problem, the `LossLayer` is just not doing what anyone would think it does (But this could be fixed with a correct docstring).

I assume that even the reduction logic of `Loss` itself is probably not well understood. It doesn't do reduction in all cases, and it has certain logic what kind of reduction it does. (To just give an idea: Remember that the loss can also be calculated inside of a recurrent layer.)

So I would say that this should better be extended for the `Loss` documentation. `LossLayer` behaves as expected when you understand what `Loss` computes.

Reply:
So for the case that I am outside a recurrent layer, is the above layer code correct to simulate what a loss would be doing if normalization is enabled? If not, it would be good if you can give some proper explanation that can be added to the documentation.

Reply:
> So for the case that I am outside a recurrent layer, is the above layer code

Which code exactly?

> correct to simulate what a loss would be doing if normalization is enabled?

Now you speak about normalization? I speak about reduction.

I explained that it does not do reduction (or it depends), and that's also what `LossLayer` returns.

The reduction is done when you ask a RETURNN network to get the total loss (must be done, obviously then, as the total loss must be a scalar, for the optimizer).

> If not, it would be good if you can give some proper explanation that can be added to the documentation.

First of all, as I explained, as I recommend to not use `LossLayer`, I'm not sure I would put it into the documentation at all.

Inside of the `LossLayer` docstring, you can simply put an example what to use instead, like what I wrote here.

Reply:
> ```
> "mse_loss": {"class": "eval", "eval": "(source(0) - source(1)) ** 2", "from": ["output", "data:classes"]}
> ```
> 
> Followed by a e.g. normalized reduce if needed:
> 
> ```
> "normalized_mse_loss": {"class": "reduce", "mode": "mean", "axis": "T", "from": "mse_loss"}
> ```
> 
> And a batch reduce to get the single loss value:
> 
> ```
> "total_mse_loss": {"class": "reduce"`, "mode": "sum", "axis": "B", "from": "normalized_mse_loss"}
> ```

This code, and it is about normalization (along the time axis) AND reduction, as usually both is done internally.
Or is it not done internally? This was my question.

Reply:
> Followed by a e.g. normalized reduce if needed

I don't understand. This is not what `LossLayer` would return, so this would be misleading.

> And a batch reduce to get the single loss value

This is not equivalent to how RETURNN would calculate the total loss. (In RETURNN, depending how you configure it, it is e.g. over all frames in the batch, e.g. via flattening over the whole batch. So doing mean over time, and then sum over batch, is sth different.)

So this would even be more misleading.

Anyway, as I explained, you should not do the reduction at all. As you see, you very easily do it in a wrong way. RETURNN will do it for you in the correct way when you use it as a loss, and otherwise, you very likely don't need it.

> Or is it not done internally

What do you mean by "internally"? As said `Loss` can do many things -- it can do reduction or not. Reduction is optional and only used under certain circumstances, like when the network calculates the total loss (`get_total_loss`). But this is a small implementation detail -- it probably would be cleaner if this code would be moved, and `Loss` would never do any reduction.

Reply:
> > Followed by a e.g. normalized reduce if needed
> 
> I don't understand. This is not what `LossLayer` would return, so this would be misleading.

This is the point of my question. I understood that the `LossLayer` is not returning the reduced loss.

> 
> > And a batch reduce to get the single loss value
> 
> This is not equivalent to how RETURNN would calculate the total loss. (In RETURNN, depending how you configure it, it is e.g. over all frames in the batch, e.g. via flattening over the whole batch. So doing mean over time, and then sum over batch, is sth different.)
> 
> So this would even be more misleading.
> 
> Anyway, as I explained, you should not do the reduction at all. As you see, you very easily do it in a wrong way. RETURNN will do it for you in the correct way when you use it as a loss, and otherwise, you very likely don't need it.
> 

But what IF I need it. All I wanted to do is to add this to the documentation of the loss (in general, not the `LossLayer`) as well. But from your response I get that this is not a good idea, and we will likely not get to a conclusion here, so lets stop at this point.

Reply:
> But what IF I need it.

Why? Very likely you should not need it. I cannot imagine any use case where you would need it. Probably you are doing it wrong. That's exactly my point here. You should not even give such an example, because this would be misleading to the user, because the user should not do the reduction (if this is used for the loss, or probably in all other cases as well). This should be done by RETURNN when needed, e.g. in `get_total_loss`.

If you give this example, it looks like this is a possible example what the user would do, and probably would copy it. But this would be wrong. The user should definitely not do it this way.

E.g. if the user now does the reduction, and then defines this as loss, he would not get what he expects.

If you want to mention the reduction, I would rather write sth like: "The user should not do the reduction, as this is done by RETURNN, when this is used for the loss. Otherwise the reduction usually does not make sense anyway."

> All I wanted to do is to add this to the documentation of the loss (in general, not the `LossLayer`) as well. But from your response I get that this is not a good idea, and we will likely not get to a conclusion here, so lets stop at this point.

I just say that the documentation should be helpful, and not be misleading, i.e. provide bad/wrong examples, or implicitly suggesting wrong layers.

Reply:
Well there is definitely a use case for this. There is an approach for TTS where you first normalize the loss along the time axis, and then exclude or clip the losses for all sequences in a batch that are outside of lets say 3 times the standard deviation of the losses in the batch.

Maybe it would be better to not use the internal loss reduction at all, but always do it explicitly if the internal reduction is too complicated or not well documented.

Reply:
> Well there is definitely a use case for this. There is an approach for TTS where you first normalize the loss along the time axis, and then exclude or clip the losses for all sequences in a batch that are outside of lets say 3 times the standard deviation of the losses in the batch.

This would be an option for the optimizer, not for the network. We already have this. It's the option `grad_norm_to_clip_to_zero`.

> Maybe it would be better to not use the internal loss reduction at all, but always do it explicitly if the internal reduction is too complicated or not well documented.

It's not complicated. It's exactly as you would expect it. But your example was wrong.

Reply:
> > Well there is definitely a use case for this. There is an approach for TTS where you first normalize the loss along the time axis, and then exclude or clip the losses for all sequences in a batch that are outside of lets say 3 times the standard deviation of the losses in the batch.
> 
> This would be an option for the optimizer, not for the network. We already have this. It's the option `grad_norm_to_clip_to_zero`.

This seems to be something completely different. But as it is not documented what ``grad`` actually is, I can not even tell exactly what it does.

> > Maybe it would be better to not use the internal loss reduction at all, but always do it explicitly if the internal reduction is too complicated or not well documented.
> 
> It's not complicated. It's exactly as you would expect it. But your example was wrong.

Yes, you told me multiple times already. But I asked you if it was correct or not, and if not I wanted to know what would be correct.

Reply:
> > > Well there is definitely a use case for this. There is an approach for TTS where you first normalize the loss along the time axis, and then exclude or clip the losses for all sequences in a batch that are outside of lets say 3 times the standard deviation of the losses in the batch.
> > 
> > This would be an option for the optimizer, not for the network. We already have this. It's the option `grad_norm_to_clip_to_zero`.
> 
> This seems to be something completely different. But as it is not documented what `grad` actually is, I can not even tell exactly what it does.

Ah sorry, yes it's not exactly the same, but almost. Why do you say totally different? It's almost the same.

What do you mean, it's not documented what grad is? I don't understand. It is very analogue to all other related options, like `grad_clip` etc.

In any case, also what you want would be an option for the optimizer. It doesn't really make sense to have this in the network.

> > > Maybe it would be better to not use the internal loss reduction at all, but always do it explicitly if the internal reduction is too complicated or not well documented.
> > 
> > It's not complicated. It's exactly as you would expect it. But your example was wrong.
> 
> Yes, you told me multiple times already. But I asked you if it was correct or not, and if not I wanted to know what would be correct.

I already wrote an example how it would be correct (in numbers, not in RETURNN layers). It's very clear, straight-forward and easy conceptually.
It is not easy to do in RETURNN layers. Mostly because there is usually never a need for that.

E.g. one way (but please do **not** put this into the documentation, as this might probably just confuse, not help) (numerically not optimal...):
```
loss: ...,  # shape [B,T,D]
lengths: {class: length, from: loss},  # shape [B]
lengths_total: {class: reduce, mode: sum, from: lengths, axes: B},  # shape []
# could also use seq_len_mask before, but not needed
loss_total_sum: {class: reduce, mode: sum, from: loss, axes: "*"},
loss_total_mean: {class: combine, from: [loss_total_sum, lengths_total], kind: real_div}
```
(The way RETURNN by default calculates it is different: It uses `flatten`. We have `UnflattenNdLayer` but not the inverse, but we could introduce such a layer. But it was never needed so far.)

Actually, I just realize, I think the `ReduceLayer` can already reduce over both B and T, like:
```
loss: ...,  # shape [B,T,D]
loss_total_mean: {class: reduce, mode: mean, axes: "*", from: loss}
```

However, even for this example, this is still not the same as what RETURNN does, because RETURNN also considers this over multiple batches (which can have different number of frames). So when you would put loss:as_is here, this is **not** the same as when you would not do the reduction. E.g. when you have one batch with 1 frame, and a next batch with 10 frames, it will calculate it correctly, but only if you don't do any reduction/normalization in your network!

To do this within the network, you would also need to keep track over the training, e.g. in a variable or some other operator with internal state. E.g. see `AccumulateMeanLayer`. But then this is also not to be used for a loss:as_is.

But this should all be quite clear, and you should know all of that.

Reply:
I only knew about the default case, so no normalization and the computation is done with `flatten_with_seq_len_mask`.

This example was actually helpful for me, I see now understand where the problems are for my specific case. And yes, it is okay for this to not be part (at least not here) of the documentation.

I now read through all the (gradient) code, and `grad_norm_to_clip_to_zero` does something completely different than what I meant from what I can see.

It sets the gradient to zero for each gradient variable if a global norm exceeds a threshold, this is not what I want. I want to exclude specific sequences from a batch (not all) from the gradient based on their individually normalized MSE loss (excluding the stop token).

This is impossible to be implemented in the optimizer. Maybe I did not explain well enough what I want to do. We can discuss this separatly.

Reply:
> I now read through all the (gradient) code, and `grad_norm_to_clip_to_zero` does something completely different than what I meant from what I can see.
> 
> It sets the gradient to zero for each gradient variable if a global norm exceeds a threshold, this is not what I want. I want to exclude specific sequences from a batch (not all) from the gradient based on their individually normalized MSE loss (excluding the stop token).

This is equivalent, i.e. excluding a sequence from a batch, or setting the gradient to zero (for a sequence from a batch).

`grad_norm_to_clip_to_zero` does that (setting gradients to zero), although for slightly different conditions (I think not per sequence, but per param, and not based on the loss, but on the gradient -- but these are details).

> This is impossible to be implemented in the optimizer. Maybe I did not explain well enough what I want to do. We can discuss this separatly.

Of course, it trivially can. The optimizer knows the loss and controls the updates. You can simply do exactly what you describe. This is the whole job of the optimizer.

Reply:
This should not be the job of the optimizer. The task of the optimizer is to define the gradients and handle the updates. It is not the task of the optimizer to define the loss. As far as I know, the optimizer only gets the total loss passed (in the default case even including the constraints, there is the flag `decouple_constraints`). The loss is passed via `TFNetwork.get_objective`, which means that different losses are hidden to the optimizer.

Options like `grad_norm_to_clip_to_zero` control the gradient for each of the variables, and work on the gradient. This has nothing to do with dropping a single sequence in a batch based on one specific loss value.